### PR TITLE
init commit action

### PR DESCRIPTION
**Pull request by fuweid**
Signed-off-by: Wei Fu <fuweid89@gmail.com>

### CODE REVIEW

File: commit.go [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+	commitCommand = &cli.Command{
+		Name:        "commit",
+		Usage:       "[flags] CONTAINER REPOSITORY[:TAG]",
+		Description: "Create a new image from a container's changes",
+		Action:      commitAction,
+		Flags: []cli.Flag{
+			&cli.StringFlag{
+				Name:    "author",
+				Aliases: []string{"a"},
+				Usage:   `Author (e.g., "nerdctl contributor <nerdctl-dev@a-team.com>")`,
```
Comment (AkihiroSuda) [inline commit.go:68]: example.com

File: commit.go [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+		RootFS: ocispec.RootFS{
+			Type:    "layers",
+			DiffIDs: append(baseConfig.RootFS.DiffIDs, diffID),
+		},
+		History: append(baseConfig.History, ocispec.History{
+			Created:    &createdTime,
+			CreatedBy:  createdBy,
+			Author:     opts.author,
+			Comment:    opts.message,
+			EmptyLayer: (diffID == emptyGZLayer),
```
Comment (AkihiroSuda) [inline commit.go:267]: This is not robust, but I can't come up with a better idea.

File: commit.go [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+
+	return ocispec.Descriptor{
+		MediaType: images.MediaTypeDockerSchema2LayerGzip,
+		Digest:    newDesc.Digest,
+		Size:      info.Size,
+	}, diffID, nil
+}
+
+// applyDiffLayer will apply diff layer content created by createDiff into the snapshotter.
+func applyDiffLayer(ctx context.Context, name string, baseImg ocispec.Image, sn snapshots.Snapshotter, differ diff.Applier, diffDesc ocispec.Descriptor) (err0 error) {
```
Comment (AkihiroSuda) [inline commit.go:371]: err0 -> retErr

**Review by AkihiroSuda — APPROVED** [review]

**Comment by AkihiroSuda** [comment]
Thanks, LGTM but a couple of nits

File: commit.go [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+	return &commitOpts{
+		author:  clicontext.String("author"),
+		message: clicontext.String("message"),
+		ref:     named.String(),
+	}, nil
+}
+
+var emptyDigest = digest.Digest("")
+
+func commitContainer(ctx context.Context, client *containerd.Client, id string, opts *commitOpts) (digest.Digest, error) {
```
Comment (AkihiroSuda) [inline commit.go:125]: Can we split this as a separate package? Like pkg/imgutil/commit

Comment (fuweid) [reply commit.go:125]: Sure. But I need to do some refactor for this. Could I handle it in the followup?

File: commit.go (hunk 2 above)
Comment (fuweid) [reply commit.go:267]: Yeah. It is kind of hacky. 😂

File: commit.go (hunk 4 above)
Comment (AkihiroSuda) [reply commit.go:125]: opened https://github.com/AkihiroSuda/nerdctl/pull/43

**Pull request merged**